Reimplement floating-point description implementation in Swift.

tbkka opened the pull request:
This replaces the previous SwiftDtoa.cpp with the same algorithm reimplemented in Swift.  It supports Float16, Float32, Float64, and Float80 (on Intel).

Performance is reasonable:  In my testing (M1 and x86_64):  Float16 and Float32 are a bit faster than the C version, Float64 is almost exactly the same, Float80 is a bit slower.

I think I've finally worked out the availability, though I'd appreciate someone who knows better taking a critical look.

tbkka committed:
Reimplement floating-point description implementation in Swift.

Note:  This is still incomplete.  I need to finish porting the Float16
and Float80 support before it can fully replace the existing
C implementation.

tbkka commented on stdlib/public/core/FloatingPointToString.swift:
```diff
@@ -0,0 +1,1291 @@
+
+@available(macOS 9999, *)
+// TODO: This doesn't guarantee inlining in all cases :(
+@inline(__always)
```
@eeckstein Why might this inlining fail in release builds?   I've seen it both succeed and fail, possibly relating to whether I have enabled some local changes that force-inline a bunch of MutableSpan APIs that are used here.  Is it possible that this is close to some inlining limit?

tbkka replied:
>  Is it possible that this is close to some inlining limit?

Perhaps related to the depth of nesting?

eeckstein replied:
> Why might this inlining fail in release builds? 
> Is it possible that this is close to some inlining limit?

yes, see https://github.com/swiftlang/swift/blob/439afd640693cf45f2d2f5f60f4c6586d7020ca9/lib/SILOptimizer/Transforms/PerformanceInliner.cpp#L932

tbkka replied:
> Why do you need `@inline(__always)` here? Is it required to overrule SIL and LLVM inlining heuristics for performance?

Thank you for that pointer!   Why is the limit 64 there?  Was there some testing that went into that choice?

For the C version, inlining this function does seem to make a big difference:  This function has a lot of cases, but only one case is used at any one time, so the cost of setting up 7 function arguments in the caller does seem to be a significant portion of the runtime.  The biggest problem so far has been that the arbitrary limits on `@inline(__always)` make it hard to experiment.  I would like to at least try inlining to see what impact it has.

eeckstein replied:
The original motivation of `@inline(__always)` was to tweak the inlining decision of some small, time-critical, stdlib functions. It was not meant to be a user-visible attribute. Nevertheless, people started to (mis-)use it, e.g. by placing it everywhere they thought it might affect performance. But in reality, this does not work because it can easily blow up the compiler. Therefore we added this quite strict limits.
`@inline(__always)` is not what it sounds like because it's name is misleading.

tbkka commented:
CC: @stephentyrone for visibility.  (I'll eventually need advice about how to deal with the availability mess here.)

tbkka committed:
Add Float16 support to the Swift implementation

tbkka committed:
Float16 tweaks: balancing safety and performance

tbkka committed:
Float32/64 safety improvements

tbkka committed:
Initial Float80 support + some general reorganization and commenting

tbkka committed:
Float80 performance

tbkka committed:
Comment updates

tbkka commented on stdlib/public/core/FloatingPointToString.swift:
```diff
@@ -0,0 +1,2408 @@
+    }
+}
+
+@available(macOS 9999, *)
```
@al45tair  I had hoped your recent work would let me eliminate these availability markings, since this is stdlib code relying on other stdlib code.  What am I missing?

al45tair replied:
Instead of removing the availability markings, you need to write e.g. `StdlibDeploymentTarget <version>`. That gets rewritten at build time to match the system we're building for, except when we're doing a release where it gets set to the same as `SwiftStdlib <version>`.

al45tair replied:
If that's the right version, yes — I saw a `macOS 15.0` somewhere else in the file so assumed 6.0 might be the right choice.

tbkka replied:
Still not there.  I changed the availability here:
```
@available(SwiftStdlib 6.2, *)
@frozen
@safe
@_addressableForDependencies
public struct InlineArray<let count: Int, Element: ~Copyable>: ~Copyable {
  @usableFromInline
  internal var _storage: Builtin.FixedArray<count, Element>
}
```
to use 
```
@available(StdlibDeploymentTarget 6.2, *)
```
and now I get
```
    `- error: values in generic types are only available in macOS 26.0.0 or newer
```

tbkka replied:
Hmmm....  Maybe @Azoy has some idea what's going on in this last case.  I guess there's somewhere that still needs to understand `StdlibDeploymentTarget`?

tbkka replied:
Oh.  Maybe that's unrelated, actually.  We still haven't landed all the new version numbers, have we?

tbkka replied:
That doesn't make sense.  Having a value in the generic arguments is a requirement on a particular compiler version to build this stdlib.  It has absolutely nothing to do with macOS version, and since all the code involved is in the same stdlib source tree, it's not a stdlib versioning issue either.  (It is very weird to say that some part of the stdlib depends on the version of some other part of the same stdlib.)

tbkka replied:
I think I worked it out.

tbkka committed:
Fix availability

tbkka committed:
Remove old C implementation

tbkka commented on stdlib/public/core/FloatingPointToString.swift:
```diff
@@ -0,0 +1,2412 @@
+  value f: Float16,
+  buffer utf8Buffer: inout MutableSpan<UTF8.CodeUnit>) -> Range<Int>
+{
+    if #available(SwiftStdlib 6.2, *) {
```
Does this necessarily do a runtime OS version check?   How does one avoid that?

In effect, we are re-implementing an old API using new ones.  The new ones are always "available" in the sense that they're always compiled -- the availability on the new APIs is only there to inform clients of their back-deployment options.  (Our availability notions seem to mistakenly conflate "when something was introduced" with "what requirements it has".)

al45tair replied:
It won't do a runtime version check _if_ the target that we're building for is high enough — it doesn't always do it.

Entirely coincidentally, I was thinking this morning  (before I'd seen this comment) that we really should have both `@available()` and `@requires()` and that we were really conflating exactly those two notions.

tbkka replied:
Perhaps we should add `@introduced()` and `@requires()` and document that `@available()` implies both of the others?  That would give a migration path.

tbkka replied:
I'm sticking with the `if #available(SwiftStdlib 6.2, *)` checks for now.  But I really think we should be able to do better.

compnerd reviewed (commented):
This seems to be missing the addition of the new file to swiftCore in the new runtimes build.

tbkka commented:
> This seems to be missing the addition of the new file to swiftCore in the new runtimes build.

How do I fix that?   I'm not familiar with "the new runtimes build."

etcwilde commented:
> How do I fix that? I'm not familiar with "the new runtimes build."

Add `FloatingPointToString.swift` to `Runtimes/Core/core/CMakeLists.txt`
https://github.com/swiftlang/swift/blob/e690f1d6c482aef2eb82c58756d31bf8c457c158/Runtimes/Core/core/CMakeLists.txt#L82

tbkka committed:
Add FP-to-string support to core library

tbkka committed:
Match up conditional for Float80 support

tbkka commented:
@swift-ci Please test

natecook1000 commented on stdlib/public/core/FloatingPointToString.swift:
```diff
@@ -0,0 +1,2413 @@
+// Support Legacy ABI on top of new implementation
+@available(SwiftStdlib 6.2, *)
+@_silgen_name("swift_float16ToString")
+public func _float16ToStringImpl(
+  _ textBuffer: UnsafeMutablePointer<UTF8.CodeUnit>,
+  _ bufferLength: UInt,
+  _ value: Float16,
+  _ debug: Bool) -> UInt64 {
+    // Code below works with raw memory.
+    var buffer = unsafe MutableSpan<UTF8.CodeUnit>(_unchecked: textBuffer,
+                                                   count: Int(bufferLength))
+    let textRange = Float16ToASCII(value: value, buffer: &buffer)
+    let textLength = textRange.upperBound - textRange.lowerBound
+
+    // Move the text to the start of the buffer
+    if textRange.lowerBound != 0 {
+        unsafe _memmove(dest: textBuffer,
+                        src: textBuffer + textRange.lowerBound,
+                        size: UInt(truncatingIfNeeded: textLength))
+    }
+    return UInt64(truncatingIfNeeded: textLength)
+}
```
This file's formatting is a bit out of sync with [standard library style](https://github.com/swiftlang/swift/blob/main/docs/StandardLibraryProgrammersManual.md#coding-style) – the main fixes I see are:
- Please use 2-space indents instead of 4
- In function calls that have to wrap their parameters, add a break before the first parameter, instead of after:
  ```swift
  var buffer = unsafe MutableSpan<UTF8.CodeUnit>(
    _unchecked: textBuffer,
    count: Int(bufferLength))
   ```
- For functions like this one, where the parameter list wraps, add a break before the closing parenthesis:
  ```swift
  public func _float16ToStringImpl(
    _ textBuffer: UnsafeMutablePointer<UTF8.CodeUnit>,
    _ bufferLength: UInt,
    _ value: Float16,
    _ debug: Bool
  ) -> UInt64 {
  ```

tbkka committed:
Try to match stdlib coding conventions

tbkka committed:
Remove a layer of indirection

tbkka committed:
Fix build when Int is 16 bits

tbkka commented:
I tweaked one place where 32-bit int was assumed but not specified.  This may let it build on AVR.

tbkka committed:
Fix Float32 formatting on 32-bit architectures

tbkka commented:
32-bit WASM tests revealed a problem with Float32 formatting on 32-bit CPUs.  There also seems to be some problem with Float64 and NaNs that I'm trying to track down.

tbkka commented:
@swift-ci Please test Linux

tbkka committed:
Remove extra semicolons

tbkka committed:
Try to fix a crash only seen so far on WASI32

tbkka committed:
Minor safety and performance tweaks

tbkka commented:
@swift-ci Please test

tbkka commented on stdlib/public/core/FloatingPointTypes.swift.gyb:
```diff
@@ -120,21 +116,37 @@ extension ${Self}: CustomDebugStringConvertible {
-    return unsafe buffer.withBytes { (bufferPtr) in
-      unsafe String._fromASCII(
-        UnsafeBufferPointer(start: bufferPtr, count: length))
+    if #available(SwiftStdlib 6.2, *) {
```
@tshortli  @al45tair  --  In my local testing, it appears that this check results in a call to `__isPlatformVersionAtLeast` which takes a full 1/3 of the run time for `debugDescription`.  This shouldn't be necessary at all -- the implementation here has no dependencies outside of the standard library that we're part of.  How can we eliminate this?

Catfish-Man replied:
Local stdlib builds set the deployment target to something way too old, see 144244211

al45tair replied:
You could also use `StdlibDeploymentTarget 6.2` here (but as before, you might end up having to move more things to `StdlibDeploymentTarget 6.2` as a result). That will get rid of the check, because when you build locally, you'll be testing for the target you're building for.

That would let you test the performance sensibly here, I think.

mikeash replied:
You can override the default deployment target when building locally by passing something like `--darwin-deployment-version-osx 26` to `--build-script`, if you just want to look at performance locally.

tbkka replied:
> You could also use StdlibDeploymentTarget 6.2 here (but as before, you might end up having to move more things to StdlibDeploymentTarget 6.2 as a result). 

Nope.  That doesn't work for `InlineArray`, at least not with a stock `build-script -rA` invocation, since "value generics are only available on macOS 26."  I discussed this in #83710.  If I landed that, then I could duplicate `InlineArray` to make `internal _InlineArray` which would not need any availability restrictions at all.  Using the same pattern to create `internal _MutableSpan` would probably just eliminate all of the availability checks in this code.

I've not experimented with `_unreachable()` yet.

I'm playing with build-script incantations to see what other options exist.

tbkka replied:
> ... something like --darwin-deployment-version-osx 26 to build-script, if you just want to look at performance locally

Sure, but that still misses the point: Nothing in this code requires macOS 26.  Everything it needs is in libswiftCore. The runtime host version is irrelevant.

mikeash replied:
It would let you look at performance locally without the version checks getting in the way. It doesn’t solve the larger issue, of course.

tbkka committed:
Prefill the work buffer with '0' characters

For example, this avoids the need to do any explicit byte-by-byte
writes when expanding "123" out to "123000000.0".

This also required reworking the "back out extra digits" process
for Float64 to ensure the unused digits get written as '0' characters
instead of null bytes.

tbkka commented on stdlib/public/core/FloatingPointTypes.swift.gyb:
```diff
@@ -120,21 +116,37 @@ extension ${Self}: CustomDebugStringConvertible {
-      unsafe String._fromASCII(
-        UnsafeBufferPointer(start: bufferPtr, count: length))
+    if #available(SwiftStdlib 6.2, *) {
+        var buffer = InlineArray<64, UTF8.CodeUnit>(repeating: 0)
```
Done.

tbkka commented:
@swift-ci Please test

tbkka commented:
@swift-ci Please test macOS Platform

tbkka committed:
Try to fix up the availability of the Swift and C projections of the legacy ABI

tbkka committed:
Adjust availability checks for float16

tbkka committed:
another attempt at fixing the float16 ABI break

tbkka committed:
Try to fix up availability of legacy shim functions

tbkka committed:
Another try at fixing the ABI change

tbkka committed:
Document false-positive API change

tbkka commented on stdlib/public/core/FloatingPointToString.swift:
```diff
@@ -0,0 +1,2613 @@
+// Support Legacy ABI on top of new implementation
+@_silgen_name("swift_float32ToString")
+@usableFromInline
+internal func _float32ToStringImpl(
```
I added the "new API" lines to stability-stdlib-abi-without-asserts.test

tbkka commented:
@swift-ci Please test macOS Platform

tbkka commented:
@swift-ci Please test Windows Platform

tbkka commented on stdlib/public/core/FloatingPointToString.swift:
```diff
@@ -0,0 +1,2408 @@
+// ================================================================
+
+// Support Legacy ABI on top of new implementation
+@_silgen_name("swift_float32ToString2")
```
It would be nice to rip out these legacy shims at some point, but that can wait.
